File: src/palette.c
#include "palette.h"
#include "boxes.h"
#include "images.h"
#include "pd_api.h"
#include "screen.h"
#include "vector.h"

#include <math.h>
#include <stdlib.h>

const vec2i paletteSize = {32, 32};

vec2f createPalettePosition(ScreenData screen);

palette createPalette(vec2f position, LCDBitmap *image, PlaydateAPI *pd) {
    LCDSprite *sprite = imagedSprite(pd, paletteSize, image);
    palette generated = {sprite, position};
    return generated;
}

void fillPalettes(palette palettes[], int quantity, ScreenData screen, LCDBitmap *image,
                  PlaydateAPI *pd) {
    for (int i = 0; i < quantity; i++) {
        vec2f position = createPalettePosition(screen);

        palette current = createPalette(position, image, pd);
        pd->sprite->setCollisionsEnabled(current.sprite, 1);

        PDRect collider = {4, 8, paletteSize.x - 8, paletteSize.y - 8};
        pd->sprite->setCollideRect(current.sprite, collider);

        palettes[i] = current;
    }
}

palette shiftPalette(palette current, float threshold, ScreenData screen,
                     LCDBitmap *image, PlaydateAPI *pd) {
    vec2f transformed = current.position;
    transformed.y--;

    if (transformed.y < threshold) {
        transformed = createPalettePosition(screen);
    }

    pd->sprite->setImage(current.sprite, image, kBitmapUnflipped);
    pd->sprite->moveTo(current.sprite, current.position.x, current.position.y);
    pd->sprite->markDirty(current.sprite);

    palette updated = {current.sprite, transformed};
    return updated;
}

vec2f createPalettePosition(ScreenData screen) {
    float xpos = (float)rand() / (float)(RAND_MAX / (screen.bounds.x));
    float ypos = (float)rand() / (float)(RAND_MAX / (screen.bounds.y));
    if (ypos <= 128)
        ypos = 128;
    vec2f originalPosition = {xpos, ypos};
    vec2f position = fenceInside(originalPosition, screen);
    return position;
}

File: Sources/PackageResolved/Palette.swift
import Charolette
import PlaydateKit

struct Palette {
    var sprite: Playdate.Sprite.Sprite
    var position: Vector2<Float>

    init(sprite: Playdate.Sprite.Sprite, position: Vector2<Float>) {
        self.sprite = sprite
        self.position = position
    }

    init(image: Playdate.Graphics.Bitmap, at position: Vector2<Float>) {
        self.sprite = Images.imagedSprite(bitmap: image, at: position)
        self.position = position
    }
}

enum Palettes {
    typealias Bitmap = Playdate.Graphics.Bitmap

    private enum Constants {
        static var paletteSize: Vector2<Float> { Vector2(x: 32, y: 32) }
    }

    static func fill(palettes: inout [Palette], screen: ScreenData, image: Bitmap) {
        for index in 0..<palettes.count {
            let position = createPalettePosition(screen: screen)
            let current = Palette(image: image, at: position)
            current.sprite.collisionsEnabled = true
            current.sprite.collideRect = .init(x: 4, y: 8, width: Constants.paletteSize.x - 8, height: Constants.paletteSize.y - 8)
            palettes[index] = current
        }
    }

    static func shift(palette: Palette, threshold: Float, screen: ScreenData, image: Bitmap) -> Palette {
        var newPosition = palette.position
        newPosition.y -= 1
        if newPosition.y < threshold { newPosition = createPalettePosition(screen: screen) }

        palette.sprite.setImage(image: image)
        palette.sprite.moveTo(x: newPosition.x, y: newPosition.y)
        palette.sprite.markDirty()
        
        return Palette(sprite: palette.sprite, position: newPosition)
    }

    private static func createPalettePosition(screen: ScreenData) -> Vector2<Float> {
        let xPosition = Float.random(in: 0...screen.bounds.x)
        let yPosition = Float.random(in: 128...screen.bounds.y)
        let originalPosition = Vector2<Float>(x: xPosition, y: yPosition)
        return screen.fencingIn(point: originalPosition)
    }
}